[reflect] sdk/log/xlog: Add FilterProcessor and EnabledParameters

pellared opened the pull request:
Per https://github.com/open-telemetry/opentelemetry-go/pull/6271#discussion_r1946251101

This introduces a new `go.opentelemetry.io/otel/sdk/log/xlog` module.
I created it because of two reasons:
- Add the possibility to filter based on the resource and scope which is available for the SDK. The scope information is the most important as it gives the possibility to e.g. filter out logs emitted for a given logger. Thus e.g. https://github.com/open-telemetry/opentelemetry-specification/issues/4364 is not necessary. See https://github.com/open-telemetry/opentelemetry-specification/pull/4290#discussion_r1927546170 for more context.
- I think we may want to stabilize the Logs SDK before https://github.com/open-telemetry/opentelemetry-specification/issues/4363  is done and stabilized. Therefore, this feature is added a separate module. `go.opentelemetry.io/otel/sdk/log` module does not expose these experimental feature in its exported API; it is only available in docs and example test. The user would have to explicitly opt-in by using the `go.opentelemetry.io/otel/sdk/log/xlog` module which will always have `v0.x.y` versions. Hopefully at some point of time the features are added to `go.opentelemetry.io/otel/sdk/log`, deprecated in `go.opentelemetry.io/otel/sdk/log/xlog` and afterwards removed from `go.opentelemetry.io/otel/sdk/log/xlog` . I think that for at one release we would need to support both the deprecated `go.opentelemetry.io/otel/sdk/xlog.FilterProcessor` and new `go.opentelemetry.io/otel/sdk/log.FilterProcessor` to facilitate the migration process. The module name is inspired by `xerrors`.`xlog` is under `sdk/log` so that it could have access to packages under `sdk/log/internal` if necessary.
- This is also related to https://github.com/open-telemetry/opentelemetry-go/issues/5882#issuecomment-2637595141
- This will also help me with working towards https://github.com/open-telemetry/opentelemetry-specification/issues/4363 as I would have a good example how we do support it.

The SDK supports `xlog` using `reflect` (duck typing) and the `sdk/log` production code does not depend on `xlog` so users should never experience build failures because of breaking changes in `xlog`.

EDIT: The performance difference is huge because of using reflect. It makes the feature even almost unusable as in most cases calling Enabled (when all Processors are FilterProcessors) would be a lot more expensive than creating a heavy log record. But maybe it  is a cost that we need to accept when providing experimental feature with a new types? The reflect overhead would be removed if the types would be stabilized and moved to `go.opentelemetry.io/otel/sdk/log`. 

```
goos: linux
goarch: amd64
pkg: go.opentelemetry.io/otel/sdk/log
cpu: 13th Gen Intel(R) Core(TM) i7-13800H
                 │   old.txt   │                  new.txt                  │
                 │   sec/op    │     sec/op      vs base                   │
LoggerEnabled-20   4.589n ± 1%   1977.500n ± 3%  +42992.18% (p=0.000 n=10)

                 │   old.txt    │            new.txt             │
                 │     B/op     │     B/op      vs base          │
LoggerEnabled-20   0.000Ki ± 0%   1.811Ki ± 6%  ? (p=0.000 n=10)

                 │  old.txt  │           new.txt            │
                 │ allocs/op │ allocs/op   vs base          │
LoggerEnabled-20   0.00 ± 0%   16.00 ± 0%  ? (p=0.000 n=10)
```

Maybe I should just use this PR as a prototype for sake of  https://github.com/open-telemetry/opentelemetry-specification/issues/4363. Then we could add the `FilterProcessor` straight to the `go.opentelemetry.io/otel/sdk/log`. However, then I have no idea how long we would need to wait for stability of Logs SDK.

The other alternative is to go with https://github.com/open-telemetry/opentelemetry-go/pull/6271 to achieve performance for the cost of coupling the `sdk/log` to `sdk/log/xlog` (similarly to what Collector does). I think the performance overhead is less acceptable than the potential build failures that users can run into. Trade-offs...

pellared commented on versions.yaml:
```diff
@@ -40,4 +40,5 @@ module-sets:
       - go.opentelemetry.io/otel/schema
 excluded-modules:
   - go.opentelemetry.io/otel/internal/tools
+  - go.opentelemetry.io/otel/sdk/log/xlog
```
I think that relying on git hashes would be very annoying for the users
 It would be hard for them to keep `sdk/log` and `sdk/log/xlog` in sync.
1. We are making commits very frequently
2. It would be hard to figure it out which version of `xlog` is compatible with `log` with we do any breaking change.

On the other side, we can wait for making a release until people actually complain.

Thoughts?

dmathieu replied:
For comparison, the collector keeps x packages within the normal release process.

I have doubts about the idea, but if we want to allow folks to be able to get very fresh code, we could setup auto nightly/weekly releases for the x packages. Then, they wouldn't rely on commits, but would remain on the latest.

pellared commented on CHANGELOG.md:
```diff
@@ -25,6 +25,11 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
+  This new module contains experimental features of the OpenTelemetry Logs SDK.
+  This module is unstable and breaking changes may be introduced.
+  See our [versioning policy](VERSIONING.md) for more information about these stability guarantees. (#6271)
+- Add `FilterProcessor` and `EnabledParameters` in `go.opentelemetry.io/otel/sdk/log/xlog` (#6271).
```
```suggestion
- Add `FilterProcessor` and `EnabledParameters` in `go.opentelemetry.io/otel/sdk/log/xlog` .
  Compared to previous version it additionally gives the possibility to filter by resource and instrumentation scope. (#6271)
```

dmathieu commented on CHANGELOG.md:
```diff
@@ -25,6 +25,13 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
+  This new module contains experimental features of the OpenTelemetry Logs SDK.
+  It replaces `go.opentelemetry.io/otel/sdk/log/internal/x`.
+  This module is unstable and breaking changes may be introduced.
+  See our [versioning policy](VERSIONING.md) for more information about these stability guarantees. (#6271)
```
```suggestion
  See our [versioning policy](VERSIONING.md) for more information about these stability guarantees. (#6286)
```

dmathieu commented on CHANGELOG.md:
```diff
@@ -25,6 +25,13 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
+  This module is unstable and breaking changes may be introduced.
+  See our [versioning policy](VERSIONING.md) for more information about these stability guarantees. (#6271)
+- Add `FilterProcessor` and `EnabledParameters` in `go.opentelemetry.io/otel/sdk/log/xlog` .
+  Compared to previous version it additionally gives the possibility to filter by resource and instrumentation scope. (#6271)
```
```suggestion
  Compared to previous version it additionally gives the possibility to filter by resource and instrumentation scope. (#6286)
```

dmathieu reviewed (commented):
Would it be possible to add a linter to ensure no stable module uses an `x` one?

dmathieu commented on sdk/log/go.mod:
```diff
@@ -10,6 +10,7 @@ require (
 	go.opentelemetry.io/otel v1.34.0
 	go.opentelemetry.io/otel/log v0.10.0
 	go.opentelemetry.io/otel/sdk v1.34.0
+	go.opentelemetry.io/otel/sdk/log/xlog v0.0.0-00010101000000-000000000000
```
This is a bit strange. I've just tried that for otelhttp, and the SDK does come as an indirect dependency.

pellared replied:
Then I think we cannot rely on the Go modules pruning and I will get rid of this dependency,.

pellared replied:
Adressed https://github.com/open-telemetry/opentelemetry-go/pull/6286/commits/08ac14607957bbeab843a5d5931546495a70e64f

pellared replied:
> This is a bit strange. I've just tried that for otelhttp, and the SDK does come as an indirect dependency.
@dmathieu , I tried it once more and actually it does NOT come as an indirect dependency.

See: https://github.com/open-telemetry/opentelemetry-go-contrib/pull/6763

I had to run `make go-mod-tidy` twice. Probably if we used `crosslink tidylist`  then it would not be necessary

We can consider reverting https://github.com/open-telemetry/opentelemetry-go/pull/6286/commits/08ac14607957bbeab843a5d5931546495a70e64f and making a rule that makes sure that production code does not use unstable modules. I can of course do it in a separate PR.

dmathieu replied:
Ah, it looks like the difference is that you used a different package name. So that'd be the trick.

pellared replied:
> Ah, it looks like the difference is that you used a different package name. So that'd be the trick.

I do not follow. The SDK is used in tests of both of `otelhttp` and `otelhttp_test` package.

dmathieu replied:
Yes, but when I tried it, I kept the `otelhttp` package name for the tests. I think that's why the dependency still showed up.

pellared replied:
Some tests still use it. Can you double check it? Make sure to run `make go-mod-tidy` more than once 😉

pellared commented:
> Would it be possible to add a linter to ensure no stable module uses an `x` one?

I created https://github.com/open-telemetry/opentelemetry-go/issues/6298 to avoid scope creeping. 
I added it to Logs GA project to make sure we address it before going GA with logs.

pellared commented:
> > Would it be possible to add a linter to ensure no stable module uses an `x` one?
> 
> I created #6298 to avoid scope creeping. I added it to Logs GA project to make sure we address it before stabilizing the logs.

I found some time and I added the rule: https://github.com/open-telemetry/opentelemetry-go/pull/6286/commits/25937e03a8a4eb8a89ae06af05b80e6fa02c9df5

pellared commented:
I think that the performance overhead is huge and that we should reconsider https://github.com/open-telemetry/opentelemetry-go/pull/6271.

pellared commented:
SIG meeting notes:
We agreed that we can move `FilterProcessor` directly to `sdk/log` as Logs SDK does not look to be stabilized soon.
